core/dbt/context/providers.py

```diff
@@ -1410,7 +1410,8 @@ def sql(self) -> Optional[str]:
         if self.model.language == ModelLanguage.sql:  # type: ignore[union-attr]
             # If the model is deferred and the adapter doesn't support zero-copy cloning, then select * from the prod
             # relation
-            if getattr(self.model, "defer_relation", None):
+            # TODO: avoid routing on args.which if possible
```

Comment:
out of scope for this PR, but one way to do this could be to have the clone task further mutate the manifest after merge_from_artifact is called to overwrite compiled_code based on defer_relation. that way we could get rid of this if block entirely

Reply:
True. The risk(?) of that approach is, the `clone` task would produce a genuinely different (compiled) manifest from other tasks, and overwrite it at the end of the task [here](https://github.com/dbt-labs/dbt-core/blob/5488dfb9926c3f81e6a80bd0d8b01fe7f0ac7a60/core/dbt/task/runnable.py#L488-L491) with potentially surprising implications for metadata consumers. (I don't think there would be any effect for `state:modified`, which doesn't use `compiled_code` at all.)

That might still be preferable to the approach in this PR.

Reply:
I'm okay to leave this as a TODO and consider alternative approaches down the line to keep risk minimal. Removing the command-based check here also feels less satisfying if downstream metadata consumers will need to put something similar in place to preserve their expected functionality.